error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/JUnitFactoryMethodDeclaration.java

```diff
@@ -0,0 +1,285 @@
+ * provides test cases for, but with a `TestCases` suffix, and - has a comment which connects the
+ * return statement to the names of the parameters in the corresponding test method.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary = "JUnit factory method declaration can likely be improved",
+    link = BUG_PATTERNS_BASE_URL + "JUnitFactoryMethodDeclaration",
+    linkType = CUSTOM,
+    severity = SUGGESTION,
+    tags = STYLE)
+public final class JUnitFactoryMethodDeclaration extends BugChecker
+    implements BugChecker.MethodTreeMatcher {
+  private static final long serialVersionUID = 1L;
+
+  private static final Matcher<MethodTree> HAS_UNMODIFIABLE_SIGNATURE =
+      anyOf(
+          annotations(AT_LEAST_ONE, isType("java.lang.Override")),
+          allOf(
+              not(hasModifier(FINAL)),
+              not(hasModifier(PRIVATE)),
+              enclosingClass(hasModifier(ABSTRACT))));
+
+  /** Instantiates a new {@link JUnitFactoryMethodDeclaration} instance. */
+  public JUnitFactoryMethodDeclaration() {}
+
+  @Override
+  public Description matchMethod(MethodTree tree, VisitorState state) {
+    if (!TEST_METHOD.matches(tree, state) || !HAS_METHOD_SOURCE.matches(tree, state)) {
+      return Description.NO_MATCH;
+    }
+
+    AnnotationTree methodSourceAnnotation =
+        ASTHelpers.getAnnotationWithSimpleName(
+            tree.getModifiers().getAnnotations(), "MethodSource");
+
+    if (methodSourceAnnotation == null) {
+      return Description.NO_MATCH;
+    }
+
+    Optional<String> factoryMethodName =
+        JUnit.extractSingleFactoryMethodName(methodSourceAnnotation);
+
+    if (factoryMethodName.isEmpty()) {
+      /* If a test has multiple factory methods, not all of them can be given the desired name. */
+      return Description.NO_MATCH;
+    }
+
+    ImmutableList<MethodTree> factoryMethods =
+        Optional.ofNullable(state.findEnclosing(ClassTree.class))
+            .map(
+                enclosingClass ->
+                    MoreASTHelpers.findMethods(enclosingClass, factoryMethodName.orElseThrow()))
+            .stream()
+            .flatMap(Collection::stream)
+            .filter(methodTree -> methodTree.getParameters().isEmpty())
+            .collect(toImmutableList());
+
+    if (factoryMethods.size() != 1) {
+      /* If we cannot reliably find the factory method, err on the side of not proposing any fixes. */
+      return Description.NO_MATCH;
+    }
+
+    ImmutableList<Description> fixes =
+        getSuggestedFixes(
+            tree,
+            state,
+            methodSourceAnnotation,
+            factoryMethodName.orElseThrow(),
+            Iterables.getOnlyElement(factoryMethods));
+
+    /* Even though we match on the test method, none of the fixes apply to it directly, so we report
+    the fixes separately using `VisitorState::reportMatch`, and return `Description.NO_MATCH`. */
```

Comment:
```suggestion
    the fixes separately using `VisitorState::reportMatch`. */
```
This is kind of implied :).